Add support for Names and ID in stats format

vdemeester opened the pull request:
This adds support to display names or id of container instead of what was provided in the request.

This keeps the default behavior (`docker stats byname` will display `byname` in the `CONTAINER` colmun and `docker stats byid` will display the id in the `CONTAINER` column) but adds two new format directive.

``` bash
$ docker stats 7c3
CONTAINER           CPU %               MEM USAGE / LIMIT       MEM %               NET I/O              BLOCK I/O             PIDS
7c3                 15.02%              666.5 MiB / 15.56 GiB   4.18%               22.46 MB / 3.56 MB   405.5 MB / 162.1 MB   71
$ docker stats --format="table {{.Names}}\t{{.ID}}\t{{.Container}}" 7c3
NAME                CONTAINER ID                                                       CONTAINER
elastic_agnesi      7c32fcc1c619123926f2c3cbc2356375b1887196b6b0cf3ed682973916bbb74e   7c3
```

/cc @thaJeztah @dnephin 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

MHBauer commented on api/types/stats.go:
```diff
@@ -170,6 +170,9 @@ type Stats struct {
 	Stats
 
+	Name string `json:"name,omitempty"`
+	ID   string `json:"id,omitempty"`
```
Why here vs on the Stats struct? I don't have any preference. I'm just curious how we decide to put it on one or the other.

vdemeester replied:
Mainly because I wanted to keep `Stats` struct with only metrics field and no metadata.

MHBauer commented on cli/command/formatter/stats.go:
```diff
@@ -25,7 +25,9 @@ const (
 type StatsEntry struct {
+	Container        string
 	Name             string
+	ID               string
```
Check my understanding.
+ Container has the user input name.
+ Name has the full name, and used to have the user input name.
+ ID has the full ID.

vdemeester replied:
yep, that's it. I kept `Container` as the user input as this is what it's doing now.

thaJeztah replied:
Could you put `ID` at the top here? (ID, Name, Container)

crosbymichael replied:
No reason to nit pick when its still in design review @thaJeztah.  be nice to @vdemeester

MHBauer reviewed (commented):
Code looks right. Not sure I understand the use case.

crosbymichael commented on cli/command/formatter/stats.go:
```diff
@@ -139,7 +141,18 @@ type containerStatsContext struct {
+	return c.s.Container
+}
+
+func (c *containerStatsContext) Names() string {
```
Why is this called `Names` with an `s`?

vdemeester replied:
@crosbymichael I asked myself the same question :sweat_smile:. I did it to *mimic* the `docker ps --format` which has `Names` too, I'm fine to rename it to `Name` 👼

cpuguy83 replied:
@vdemeester `docker ps` used to show all the things a container could be referenced as which includes names, as well as link aliases. That's why it is plural.... it might still do that if you use `--no-trunc` come to think of it.

Definitely singular here is good.

vdemeester replied:
I'll update 👼

ripcurld00d commented on cli/command/container/stats_helpers.go:
```diff
@@ -46,15 +46,15 @@ func (s *stats) remove(id string) {
 
 func collect(s *formatter.ContainerStats, ctx context.Context, cli client.APIClient, streamStats bool, waitFirst *sync.WaitGroup) {
-	logrus.Debugf("collecting stats for %s", s.Name)
+	logrus.Debugf("collecting stats for %s", s.Container)
```
Can't we display both the container name and the container ID?
I believe the container name is more readable than its iD.
Thought? :blush:

vdemeester replied:
@ripcurld00d On the debug log ? This change is here to keep the same behavior.

ripcurld00d replied:
@vdemeester yes, now I see it. s.Container can be the ID or the Name depends on what the user enters.

ripcurld00d commented:
@vdemeester @thaJeztah 
I think it closes #20973, right?

ripcurld00d commented:
If I'm not wrong, in case there is a container called `18f132d776` and another container with ID `18f132d776` then both of them will be displayed.

Is that OK? Maybe adding `--container-id` can solve this edge-case?
Feedback? :smiley_cat:

vdemeester commented:
@ripcurld00d yes you're right, I should update the description. For the name/id collision, I don't think we should do anything 👼.

coolljt0725 commented:
During review this PR, I found there would be an issue with `docker stats byname` if user has call `docker rename ` to rename the container, the name display in ` CONTAINER` colmun does not change.  I think this PR could fix this issue with more changes since this PR return the container name to client(https://github.com/docker/docker/pull/27797/files#diff-ad75d4bbd2897e0abd0207e0f2608bf6R47).

vdemeester commented:
@coolljt0725 yep, I'm not really sure how to fix it the best way possible 👼. At some point we should make it so that `Container` and `ID` are the same (or `Name`), and thus not display the user input as `Container`. But not sure how we should change that (in terms or retro-compat, etc…)…

vdemeester committed:
Add support for Names and ID in stats format

This adds support to display names or id of container instead of what
was provided in the request.

This keeps the default behavior (`docker stats byname` will display
`byname` in the `CONTAINER` colmun and `docker stats byid` will display
the id in the `CONTAINER` column) but adds two new format directive.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

crosbymichael commented:
What is the use case for this because adding three different ways have identifying the container is pretty confusing for me and hard to reason about that cons for this.

vdemeester commented:
@crosbymichael The use case is to be able to display the actual name or id of the container instead of the user input. Right now there is no way to get the name of a container for `stats` if you don't already know it and pass it to the command (and thus, `stats` on all containers, you can't, at all :sweat:).

Keeping `Container` for retro-compatibility but I really wish I wouldn't have too…

crosbymichael commented:
@vdemeester is this just an api issue or does the cli format have to change as well? Could we get away with just returning the full container id and not this name stuff since names can change and its weird.  Most of the time the name will be the `Continaer` part anyways right?

colinmollenhour commented:
This shows merged into 1.13 milestone but it does not appear to be present in 17.03.0-ce. Also, shouldn't "byname" and "byid" be "--byname" and "--byid" since they are options?
